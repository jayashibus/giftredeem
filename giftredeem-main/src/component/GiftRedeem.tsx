import React, { useState, useEffect } from "react";
import { addRedemption, getRedemption } from "../http/HttpService";
import { CONSTANTS } from '../constants'

type IStaffDetails = {
  team_name: string;
  staff_pass_id: string;
  redeemed_at: string;
}


const GiftRedeem = ({ data }: any) => {
  const [staffId, setStaffId] = useState<string>("");
  const [error, setError] = useState<string>("");
  const [validStaff, setValidStaff] = useState<boolean>(false);
  const [staffDetails, setStaffDetails] = useState<IStaffDetails>({} as any);
  const [redemptionSuccess, setRedemptionSuccess] = useState<boolean>(false);
  const [validRedemption, setValidRedemption] = useState<boolean>(true);
  const [redemptionData, setRedemptionData] = useState<Array<any>>([]);

  useEffect(() => {
    setRedemptionData([...redemptionData, ...data.redemptionData]);
  }, [data]);

  const onClickVerifyHandler = (e: any) => {
    e.preventDefault();
    if (!staffId) {
      setError(CONSTANTS.giftRedeemScreen.validationMsg.emptyFied);
    } else {
      const valid = data?.staffData.find(
        (item: any) => item.staff_pass_id === staffId
      );

      if (valid) {
        setStaffDetails(valid);
        setValidStaff(true);
        setError("");
        const alreadyRedeemed = redemptionData.find(
          (team: any) => team.team_name === valid.team_name
        );

        if (alreadyRedeemed) {
          setValidRedemption(false);
        }
      } else {
        setError(CONSTANTS.giftRedeemScreen.validationMsg.invalidStaffId);
      }
    }
  };

  const onClickRedeemHandler = (teamName: string, staffPassId: string) => {
    const currentDateTime = Date.parse(Date());
    addRedemption({
      team_name: teamName,
      staff_pass_id: staffPassId,
      redeemed_at: currentDateTime,
    });

    getRedemption().then((redemptionData: any) => {
      console.log(data);
      setRedemptionData(redemptionData);
    });

    setRedemptionSuccess(true);
  };

  const onClickHomeHandler = () => {
    setValidStaff(false);
    setRedemptionSuccess(false);
    setValidRedemption(true);
    setStaffId("");
  };

  type IMessageBox = {
    title: string;
    message: string;
  }

  const MessageBox = ({ title, message }: IMessageBox) => {
    return (
      <>
        <h1>Welcome, {title} </h1>
        <p>{message}</p>
      </>
    );
  };

  const form = () => {
    return (
      <form onSubmit={onClickVerifyHandler}>
        {validStaff ? (
          <>
            <input type="text" value={`ID : ${staffId}`} disabled />
            <input
              type="text"
              value={`Team : ${staffDetails.team_name}`}
              disabled
            />

            <span className="primary_color">
              {"Account Verified Successfully"}
            </span>
          </>
        ) : (
          <>
            <input
              type="text"
              value={staffId}
              placeholder="Enter Staff ID"
              onChange={(e) => setStaffId(e.target.value)}
            />
            {error ? <span className="error_message">{error}</span> : ""}
            <button>Verify</button>
          </>
        )}
      </form>
    );
  };

  return (
    <div className="login-page">
      <h2>{CONSTANTS.giftRedeemScreen.title}</h2>
      <div className="container" id="container">
        <div className="form-container sign-in-container">{form()}</div>
        <div className="overlay-container">
          <div
            className={`overlay${validStaff && validRedemption ? "_green" : ""
              }`}
          >
            <div className="overlay-panel overlay-right">
              {validStaff && validRedemption
                ? redemptionSuccess
                  ? (
                    <MessageBox title={staffDetails.team_name} message={CONSTANTS.giftRedeemScreen.message.success} />
                  )
                  : <MessageBox title={staffDetails.team_name} message={CONSTANTS.giftRedeemScreen.message.available} />
                : validStaff && !validRedemption
                  ? <MessageBox title={staffDetails.team_name} message={CONSTANTS.giftRedeemScreen.message.already} />
                  : <MessageBox title={CONSTANTS.giftRedeemScreen.messageTitle} message={CONSTANTS.giftRedeemScreen.message.season} />}

              {validStaff && !redemptionSuccess && validRedemption && (
                <button
                  className="success"
                  id="signUp"
                  onClick={() =>
                    onClickRedeemHandler(
                      staffDetails.team_name,
                      staffDetails.staff_pass_id
                    )
                  }
                >
                  {CONSTANTS.giftRedeemScreen.button.redeemGift}
                </button>
              )}

              {validStaff && redemptionSuccess && (
                <button
                  className="success"
                  id="signUp"
                  onClick={onClickHomeHandler}
                >
                  {CONSTANTS.giftRedeemScreen.button.serveNewteam}
                </button>
              )}

              {!validRedemption && (
                <button
                  className="success"
                  id="signUp"
                  onClick={onClickHomeHandler}
                >
                 {CONSTANTS.giftRedeemScreen.button.serveNewteam}
                </button>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default GiftRedeem;

// STAFF_H123804820G
// MANAGER_T999888420B
// BOSS_T000000001P
