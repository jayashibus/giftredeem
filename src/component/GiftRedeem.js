import React, { useState, useEffect } from "react";
import { addRedemption, getRedemption } from "./../http/HttpService";

const GiftRedeem = ({ data }) => {
  const [staffId, setStaffId] = useState("");
  const [error, setError] = useState("");
  const [validStaff, setValidStaff] = useState(false);
  const [staffDetails, setStaffDetails] = useState([]);
  const [redemptionSuccess, setRedemptionSuccess] = useState(false);
  const [validRedemption, setValidRedemption] = useState(true);
  const [redemptionData, setRedemptionData] = useState([]);

  const message = {
    success:
      "You are Successfully redeem your team Gift. Thank you. Happy Christmas..!",
    available:
      "Christmas Gifts are available for your Team. Kindly show your staff id and collect your team gifts.",
    already:
      "Already one of your team representative collected your Team gifts. Thank you. Happy Christmas.!",
    season:
      "Christmas season begins. Connected with us and send your representative to redeem your Team gift's. Thanks you.",
  };

  useEffect(() => {
    setRedemptionData([...redemptionData, ...data.redemptionData]);
  }, [data]);

  const onClickVerifyHandler = (e) => {
    e.preventDefault();
    if (!staffId) {
      setError("Staff field empty. Enter valid staff ID");
    } else {
      const valid = data.staffData.find(
        (item) => item.staff_pass_id === staffId
      );

      if (valid) {
        setStaffDetails(valid);
        setValidStaff(true);
        setError("");
        const alreadyRedeemed = redemptionData.find(
          (team) => team.team_name === valid.team_name
        );

        if (alreadyRedeemed) {
          setValidRedemption(false);
        }
      } else {
        setError("Invalid Staff ID");
      }
    }
  };

  const onClickRedeemHandler = (teamName, staffPassId) => {
    const currentDateTime = Date.parse(Date());
    addRedemption({
      team_name: teamName,
      staff_pass_id: staffPassId,
      redeemed_at: currentDateTime,
    });

    getRedemption().then((redemptionData) => {
      console.log(data);
      setRedemptionData(redemptionData);
    });

    setRedemptionSuccess(true);
  };

  const onClickHomeHandler = () => {
    setValidStaff(false);
    setRedemptionSuccess(false);
    setValidRedemption(true);
    setStaffId("");
  };

  const messageBox = (title, message) => {
    return (
      <>
        <h1>Welcome, {title} </h1>
        <p>{message}</p>
      </>
    );
  };

  const form = () => {
    return (
      <form onSubmit={onClickVerifyHandler}>
        {validStaff ? (
          <>
            <input type="text" value={`ID : ${staffId}`} disabled />
            <input
              type="text"
              value={`Team : ${staffDetails.team_name}`}
              disabled
            />

            <span className="primary_color">
              {"Account Verified Successfully"}
            </span>
          </>
        ) : (
          <>
            <input
              type="text"
              value={staffId}
              placeholder="Enter Staff ID"
              onChange={(e) => setStaffId(e.target.value)}
            />
            {error ? <span className="error_message">{error}</span> : ""}
            <button>Verify</button>
          </>
        )}
      </form>
    );
  };

  return (
    <div className="login-page">
      <h2>Welcome to our gift's redemption counter</h2>
      <div className="container" id="container">
        <div className="form-container sign-in-container">{form()}</div>
        <div className="overlay-container">
          <div
            className={`overlay${
              validStaff && validRedemption ? "_green" : ""
            }`}
          >
            <div className="overlay-panel overlay-right">
              {validStaff && validRedemption
                ? redemptionSuccess
                  ? messageBox(staffDetails.team_name, message.success)
                  : messageBox(staffDetails.team_name, message.available)
                : validStaff && !validRedemption
                ? messageBox(staffDetails.team_name, message.already)
                : messageBox("Teams!", message.season)}

              {validStaff && !redemptionSuccess && validRedemption && (
                <button
                  className="success"
                  id="signUp"
                  onClick={() =>
                    onClickRedeemHandler(
                      staffDetails.team_name,
                      staffDetails.staff_pass_id
                    )
                  }
                >
                  Redeem Gift
                </button>
              )}

              {validStaff && redemptionSuccess && (
                <button
                  className="success"
                  id="signUp"
                  onClick={onClickHomeHandler}
                >
                  Serve New Team
                </button>
              )}

              {!validRedemption && (
                <button
                  className="success"
                  id="signUp"
                  onClick={onClickHomeHandler}
                >
                  Serve New Team
                </button>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default GiftRedeem;

// STAFF_H123804820G
// MANAGER_T999888420B
// BOSS_T000000001P
