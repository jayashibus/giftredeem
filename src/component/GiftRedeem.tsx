import React, { useState, useEffect } from "react";
import { addRedemption, getRedemption } from "../service/HttpService";
import { CONSTANTS } from "../constants";
import MessageBox from "./MessageBox";
import Form from "./Form";

type IStaffDetails = {
  team_name: string;
  staff_pass_id: string;
  created_at: string;
};
export const findStaffRecord = (arr: IStaffDetails[], staffid: any) =>
  arr.find((item: IStaffDetails) => item.staff_pass_id === staffid);

const GiftRedeem = ({ data }: any) => {
  const [staffId, setStaffId] = useState<string>("");
  const [error, setError] = useState<string>("");
  const [validStaff, setValidStaff] = useState<boolean>(false);
  const [staffDetails, setStaffDetails] = useState<IStaffDetails>({} as any);
  const [redemptionSuccess, setRedemptionSuccess] = useState<boolean>(false);
  const [validRedemption, setValidRedemption] = useState<boolean>(true);
  const [redemptionData, setRedemptionData] = useState<Array<any>>([]);

  // useEffect(() => {
  //   setRedemptionData([...redemptionData, ...data.redemptionData]);
  // }, [data]);

  const onClickVerifyHandler = (e: any) => {
    e.preventDefault();

    console.log("Onclick Satffid ", staffId);

    if (!staffId) {
      setError(CONSTANTS.giftRedeemScreen.validationMsg.emptyFied);
    } else {
      const checkValidStaff: any = findStaffRecord(data.staffData, staffId);

      if (checkValidStaff) {
        setStaffDetails(checkValidStaff);
        setValidStaff(true);
        setError("");
        const alreadyRedeemed = redemptionData.find(
          (team: any) => team.team_name === checkValidStaff.team_name
        );

        if (alreadyRedeemed) {
          setValidRedemption(false);
        }
      } else {
        setError(CONSTANTS.giftRedeemScreen.validationMsg.invalidStaffId);
      }
    }
  };

  const onClickRedeemHandler = (teamName: string, staffPassId: string) => {
    const currentDateTime = Date.parse(Date());
    addRedemption({
      team_name: teamName,
      staff_pass_id: staffPassId,
      redeemed_at: currentDateTime,
    });

    getRedemption().then((redemptionData: any) => {
      console.log(data);
      setRedemptionData(redemptionData);
    });

    setRedemptionSuccess(true);
  };

  const onClickHomeHandler = () => {
    setValidStaff(false);
    setRedemptionSuccess(false);
    setValidRedemption(true);
    setStaffId("");
  };

  return (
    <div className="login-page">
      <h2>{CONSTANTS.giftRedeemScreen.title}</h2>
      <div className="container" id="container">
        <div className="form-container sign-in-container">
          {
            <Form
              validStaff={validStaff}
              staffId={staffId}
              team_name={staffDetails.team_name}
              error={error}
              onSubmit={onClickVerifyHandler}
              onChange={(e) => setStaffId(e.target.value)}
            />
          }
        </div>
        <div className="overlay-container">
          <div
            className={`overlay${
              validStaff && validRedemption ? "-green" : ""
            }`}
          >
            <div className="overlay-panel overlay-right">
              {validStaff && validRedemption ? (
                redemptionSuccess ? (
                  <MessageBox
                    title={staffDetails.team_name}
                    message={CONSTANTS.giftRedeemScreen.message.success}
                  />
                ) : (
                  <MessageBox
                    title={staffDetails.team_name}
                    message={CONSTANTS.giftRedeemScreen.message.available}
                  />
                )
              ) : validStaff && !validRedemption ? (
                <MessageBox
                  title={staffDetails.team_name}
                  message={CONSTANTS.giftRedeemScreen.message.already}
                />
              ) : (
                <MessageBox
                  title={CONSTANTS.giftRedeemScreen.messageTitle}
                  message={CONSTANTS.giftRedeemScreen.message.season}
                />
              )}

              {validStaff && !redemptionSuccess && validRedemption && (
                <button
                  className="success"
                  id="signUp"
                  onClick={() =>
                    onClickRedeemHandler(
                      staffDetails.team_name,
                      staffDetails.staff_pass_id
                    )
                  }
                >
                  {CONSTANTS.giftRedeemScreen.button.redeemGift}
                </button>
              )}

              {validStaff && redemptionSuccess && (
                <button
                  className="success"
                  id="signUp"
                  onClick={onClickHomeHandler}
                >
                  {CONSTANTS.giftRedeemScreen.button.serveNewteam}
                </button>
              )}

              {!validRedemption && (
                <button
                  className="success"
                  id="signUp"
                  onClick={onClickHomeHandler}
                >
                  {CONSTANTS.giftRedeemScreen.button.serveNewteam}
                </button>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default GiftRedeem;

// STAFF_H123804820G
// MANAGER_T999888420B
// BOSS_T000000001P
