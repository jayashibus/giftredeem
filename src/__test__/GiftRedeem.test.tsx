import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";

import GiftRedeem, { findStaffRecord } from "../component/GiftRedeem";

const data = [
  {
    staff_pass_id: "STAFF_H123804820G",
    team_name: "BASS",
    created_at: "1623772799000",
  },
  {
    staff_pass_id: "MANAGER_H123804820G",
    team_name: "RUST",
    created_at: "1623772799000",
  },
  {
    staff_pass_id: "BOSS_H123804820G",
    team_name: "RUST",
    created_at: "1623772799000",
  },
];
describe("<GiftRedeem />", () => {
  test("renders component", () => {
    const { container } = render(<GiftRedeem data={data} />);
    expect(container).toMatchSnapshot();
  });
  test("Should render buttons", () => {
    render(<GiftRedeem data={data} />);

    expect(
      screen.getByText(
        "Christmas season begins. Connected with us and send your representative to redeem your Team gift's. Thanks you."
      )
    ).toBeInTheDocument;

    const submitButton = expect(
      screen.findByRole("button", { name: /Verify/i })
    );
    const redeemButton = expect(
      screen.findByRole("button", { name: /Redeem Gift/i })
    );
    const serveNewButton = expect(
      screen.findByRole("button", { name: /Serve New Team/i })
    );
  });
});

describe("Find the Staff Account", () => {
  test("Find staff details", () => {
    expect(findStaffRecord(data, "STAFF_H123804820GSS")).toBeTruthy();
  });
});
