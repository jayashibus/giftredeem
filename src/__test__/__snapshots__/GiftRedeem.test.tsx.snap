// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiftRedeem /> renders component 1`] = `
<div>
  <div
    class="login-page"
  >
    <h2>
      Welcome to our gift's redemption counter
    </h2>
    <div
      class="container"
      id="container"
    >
      <div
        class="form-container sign-in-container"
      >
        <form>
          <input
            placeholder="Enter Staff ID"
            type="text"
            value=""
          />
          <button>
            Verify
          </button>
        </form>
      </div>
      <div
        class="overlay-container"
      >
        <div
          class="overlay"
        >
          <div
            class="overlay-panel overlay-right"
          >
            <h1>
              Welcome, 
              Teams!
               
            </h1>
            <p>
              Christmas season begins. Connected with us and send your representative to redeem your Team gift's. Thanks you.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
